src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/bucket/BucketCalculator.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.bucket;
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.Resource;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.grpc.ResourceEnum;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.store.rca.cluster.NodeKey;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A BucketCalculator identifies which {@link UsageBucket} a {@link Resource} should be placed
+ * into given that resource's value. It does this on a per-node basis.
+ *
+ * <p>E.g. a BucketCalculator can compute a value of HEALTHY for CPU on Node "A" and a value of
+ * HEALTHY_WITH_BUFFER for CPU on Node "B". A consumer of this information can then read out these
+ * bucket values by calling something like getUsageBucket(NodeA, CPU).
+ */
+public abstract class BucketCalculator {
+  private final Logger LOG = LogManager.getLogger(getClass());
+  // Strings which map to threshold definitions in rca.conf
+  public static final String YOUNG_GEN_PROMOTION_RATE = "young-gen-heap-promotion-rate";
+  public static final String OLD_GEN_HEAP_USAGE = "old-gen-heap-usage";
+  public static final String CPU_USAGE = "cpu-usage";
+
+  // For each Node, store a map of resource name to the UsageBucket for that resource
+  protected Map<NodeKey, Map<ResourceEnum, UsageBucket>> resourceBucketMap = new HashMap<>();
+
+  public void computeBucket(NodeKey nodeKey, ResourceEnum resource, double value) {
```

Comment:
Does `BucketCalculator` need to store what was the computed bucket for each resource on each node? How about we just let ClusterLevelRCAs store this info, and only use BucketCalculators to compute what the appropriate bucket is?
Right now the computed state for an RCA is spread across too many classes, which in themselves have multi-level abstract and concrete classes. It's getting hard to track.
I've not gotten to the end of the PR, doing it this way may be the best approach. But adding a note for you to reconsider if any changes can make things easier to follow.